shared/regex/codeql/regex/nfa/NfaUtils.qll

```diff
@@ -0,0 +1,1382 @@
+              .charAt(_)
+      )
+  }
+
+  /**
+   * Holds if `t` matches at least an epsilon symbol.
+   *
+   * That is, this term does not restrict the language of the enclosing regular expression.
+   *
+   * This is implemented as an under-approximation, and this predicate does not hold for sub-patterns in particular.
+   */
+  predicate matchesEpsilon(RegExpTerm t) {
+    t instanceof RegExpStar
+    or
+    t instanceof RegExpOpt
+    or
+    t.(RegExpRange).getLowerBound() = 0
+    or
+    exists(RegExpTerm child |
+      child = t.getAChild() and
+      matchesEpsilon(child)
+    |
+      t instanceof RegExpAlt or
+      t instanceof RegExpGroup or
+      t instanceof RegExpPlus or
+      t instanceof RegExpRange
+    )
+    or
+    matchesEpsilon(t.(RegExpBackRef).getGroup())
+    or
+    forex(RegExpTerm child | child = t.(RegExpSequence).getAChild() | matchesEpsilon(child))
+  }
+
+  /**
+   * A lookahead/lookbehind that matches the empty string.
+   */
+  class EmptyPositiveSubPattern instanceof RegExpSubPattern {
+    EmptyPositiveSubPattern() {
+      (
+        this instanceof RegExpPositiveLookahead
+        or
+        this instanceof RegExpPositiveLookbehind
+      ) and
+      matchesEpsilon(this.getOperand())
+    }
+
+    /** Gets a string representation of this sub-pattern. */
+    string toString() { result = super.toString() }
+  }
+
+  /** DEPRECATED: Use `EmptyPositiveSubPattern` instead. */
+  deprecated class EmptyPositiveSubPatttern = EmptyPositiveSubPattern;
+
+  /**
+   * A branch in a disjunction that is the root node in a literal, or a literal
+   * whose root node is not a disjunction.
+   */
+  class RegExpRoot instanceof RegExpTerm {
+    RegExpRoot() {
+      exists(RegExpParent parent |
+        exists(RegExpAlt alt |
+          alt.isRootTerm() and
+          this = alt.getAChild() and
+          parent = alt.getParent()
+        )
+        or
+        this.isRootTerm() and
+        not this instanceof RegExpAlt and
+        parent = this.getParent()
+      )
+    }
+
+    /**
+     * Holds if this root term is relevant to the ReDoS analysis.
+     */
+    predicate isRelevant() {
+      // is actually used as a RegExp
+      super.isUsedAsRegExp() and
+      // not excluded for library specific reasons
+      not isExcluded(super.getRootTerm().getParent())
+    }
+
+    /** Gets a string representation of this root term. */
+    string toString() { result = this.(RegExpTerm).toString() }
+
+    /** Gets the outermost term of this regular expression. */
+    RegExpTerm getRootTerm() { result = super.getRootTerm() }
+  }
+
+  /**
+   * A constant in a regular expression that represents valid Unicode character(s).
+   */
+  private class RegexpCharacterConstant instanceof RegExpConstant {
+    RegexpCharacterConstant() { this.isCharacter() }
+
+    string toString() { result = this.(RegExpConstant).toString() }
+
+    RegExpTerm getRootTerm() { result = this.(RegExpConstant).getRootTerm() }
```

Comment:
>  not sure if it can be avoided though.

It can't be avoided with the current structure of the code.  
it's not possible to `extend` a parameter in a parameterised module, so I have to use `instanceof`, which doesn't "forward" the member-predicates.  

I might have written the code using less classes (and more predicates) if I had done it as a parameterised module from the start, but with this change I can change as little as possible compared to the existing implementation.